lib/puppet/resource_api.rb

```diff
@@ -209,6 +211,27 @@ def my_provider
         raise 'Execution encountered an error' if context.failed?
       end
 
+      def strict_check(current_state)
+        return if Puppet.settings[:strict] == :off
+
+        # if strict checking is on we must notify if the values are changed by canonicalize
+        # make a deep copy to perform the operation on and to compare against later
+        state_clone = Marshal.load(Marshal.dump(current_state))
+        state_clone = my_provider.canonicalize(context, [state_clone]).first
+
+        # compare the clone against the current state to see if changes have been made by canonicalize
+        return unless state_clone && (current_state != state_clone)
+        message = "Canonicalization will modify values from #{current_state.inspect} to #{state_clone.inspect}"
+
+        case Puppet.settings[:strict]
+        when :warning
+          Puppet.warning(message)
+        when :error
+          # TODO: get more context for the abort error message
```

Comment:
What's the current output of, say, `puppet resource` when this is hit, and what should it be?